### PR TITLE
*: use fxhash

### LINKED ISSUE #3032: Improve performance of decoding table row
Currently we are heavily depends of HashSet and HashMap while decoding a row(cut_row), which do some unnecessary allocation and hash computing. Since the key is always i64/u64, and the element count is small. we can implement a highly optimize/customize a HashSet/HashMap with a internal vector.

### PR DESCRIPTION
**Pull request by BusyJay**
Most of use cases of hashing in our codebase is to hash a u64/i64. fxhash is much faster than fnvhash in this case.

A simple benchmark shows that:
```
test bench_fnv_3  ... bench:           7 ns/iter (+/- 0)
test bench_fnv_5  ... bench:           7 ns/iter (+/- 0)
test bench_fx_3   ... bench:           3 ns/iter (+/- 0)
test bench_fx_5   ... bench:           3 ns/iter (+/- 0)
test bench_iter_3 ... bench:           5 ns/iter (+/- 0)
test bench_iter_5 ... bench:           6 ns/iter (+/- 0)
```

`bench_fnv_3` means querying elements of FnvHashMap which contains 3 elements. `bench_iter_3` means using pure iteration.

close #3032, #2851 and #1843.

### CODE REVIEW

**Comment by BusyJay** [comment]
What do you mean by "only for fxHash"? As the pull request's description says,
> `bench_iter_3` means using pure iteration

.

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -378,8 +378,8 @@ impl Peer {
             proposals: Default::default(),
             apply_proposals: vec![],
             pending_reads: Default::default(),
-            peer_cache: RefCell::new(FlatMap::default()),
-            peer_heartbeats: FlatMap::default(),
+            peer_cache: RefCell::new(HashMap::default()),
```
Comment (Hoverbear) [inline src/raftstore/store/peer.rs]: So here we are supposed to be using `fxhash`? `::default()` is enough? `::with_hasher()` is not needed for this?

Comment (BusyJay) [reply src/raftstore/store/peer.rs]: `with_hasher` is not necessary. `HashMap<K, S>` is an alias of `FxHashMap<K, S>`, which is an alias of `std::collections::HashMap<K, V, FxBuildHasher>`.

Comment (Hoverbear) [reply src/raftstore/store/peer.rs]: Ah, right, I see the alias now: https://github.com/pingcap/tikv/pull/3137/files#diff-91c2e21eb3bb3e6831f5ec0cc61e41eaL15

**Pull request merged**